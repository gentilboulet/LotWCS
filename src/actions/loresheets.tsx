import * as constants from '../constants/loresheets';
import { ICost } from '../types/costs';

export interface ILoresheetOpen {
  type: constants.LORESHEET_OPEN;
  uid: string;
  cost: ICost;
}

export interface ILoresheetBuyOption {
  type: constants.LORESHEET_BUY_OPTION;
  lsUid: string;
  uid: string;
  cost: ICost;
}

export type ILoresheetAction = ILoresheetOpen | ILoresheetBuyOption;

export function open(uid: string, cost: ICost): ILoresheetOpen {
  return {
    type: constants.LORESHEET_OPEN,
    uid: uid,
    cost: cost,
  };
}

export function buyOption(lsUid: string, uid: string, cost: ICost): ILoresheetBuyOption {
  return {
    type: constants.LORESHEET_BUY_OPTION,
    lsUid: lsUid,
    uid: uid,
    cost: cost,
  };
}
