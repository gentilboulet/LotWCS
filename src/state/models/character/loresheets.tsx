import {
  getLoresheetOptionData,
  IDataLoresheetOption,
  IDataLoresheetOptionPrerequisite,
  isRepeatable,
  validateLoresheet,
  validateLoresheetOption,
} from "../../../data/loresheets";
import { ICharacterState } from "./index";

export interface ILoresheetOptionState {
  uid: string;
  payload: any;
}

export interface ILoresheetsState {
  [lsUid: string]: ILoresheetOptionState[];
}

export function createState(): ILoresheetsState {
  return {};
}

export function isLoresheetPresent(
  state: ILoresheetsState,
  loresheetUid: string,
): boolean {
  const loresheetIndex = Object.keys(state).findIndex(
    stateLoresheetUid => stateLoresheetUid === loresheetUid,
  );
  return loresheetIndex !== -1;
}

export function isLoresheetOptionPresent(
  state: ILoresheetsState,
  loresheetUid: string,
  optionUid: string,
): boolean {
  if (state[loresheetUid] === undefined) {
    return false;
  }
  const optionIndex = state[loresheetUid].findIndex(
    option => option.uid === optionUid,
  );
  return optionIndex !== -1;
}

export function canOpenLoresheet(state: ICharacterState, uid: string): boolean {
  return !isLoresheetPresent(state.loresheets, uid);
}

function _or(a: boolean, b: boolean) {
  return a || b;
}
function _and(a: boolean, b: boolean) {
  return a && b;
}

function _prerequisiteToBool(
  state: ILoresheetsState,
  lsUid: string,
  prerequisite: IDataLoresheetOptionPrerequisite,
): boolean {
  if (typeof prerequisite === "string") {
    return isLoresheetOptionPresent(state, lsUid, prerequisite);
  } else {
    return prerequisite.prerequisites
      .map((optUid: string) => {
        const here = isLoresheetOptionPresent(state, lsUid, optUid);
        return here;
      })
      .reduce(
        prerequisite.type === "OR" ? _or : _and,
        prerequisite.type === "OR" ? false : true,
      );
  }
}

export function canBuyLoresheetOption(
  state: ILoresheetsState,
  lsUid: string,
  uid: string,
): boolean {
  if (!isLoresheetPresent(state, lsUid)) {
    return false;
  } // Not opened
  const lsOptData = getLoresheetOptionData(lsUid, uid);
  const isPresent = isLoresheetOptionPresent(state, lsUid, uid);
  if (isPresent) {
    if (!isRepeatable(lsUid)) {
      return false;
    } else {
      return (lsOptData as IDataLoresheetOption).repeatable;
    }
  } // Already bought
  const cb = lsOptData.prerequisites
    .map(p => _prerequisiteToBool(state, lsUid, p))
    .reduce(_and, true);
  return cb;
}

export function buyLoresheetOption(
  state: ILoresheetsState,
  loresheetUid: string,
  optionUid: string,
  payload?: string,
): void {
  validateLoresheet(loresheetUid);
  validateLoresheetOption(loresheetUid, optionUid);

  if (!canBuyLoresheetOption(state, loresheetUid, optionUid)) {
    throw new Error(
      "Internal error : impossible to buy loresheet option " +
        loresheetUid +
        "," +
        optionUid,
    );
  }
  state[loresheetUid].push({ uid: optionUid, payload });
}

export function openLoresheet(
  state: ILoresheetsState,
  loresheetUid: string,
): void {
  validateLoresheet(loresheetUid);
  state[loresheetUid] = [];
}
