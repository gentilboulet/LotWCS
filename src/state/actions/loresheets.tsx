import * as constants from 'state/constants/loresheets';
import { ICost } from 'costs/types';
import * as dataLoresheets from 'data/loresheets';

export interface ILoresheetOpen {
  type: constants.LORESHEET_OPEN;
  uid: string;
  cost: ICost;
}

export interface ILoresheetBuyOption {
  type: constants.LORESHEET_BUY_OPTION;
  lsUid: string;
  uid: string;
  cost: ICost;
}

export type ILoresheetAction = ILoresheetOpen | ILoresheetBuyOption;

export function open(uid: string, cost: ICost): ILoresheetOpen {
  dataLoresheets.validateLoresheet(uid);
  return { cost, type: constants.LORESHEET_OPEN, uid };
}

export function buyOption(lsUid: string, uid: string, cost: ICost): ILoresheetBuyOption {
  dataLoresheets.validateLoresheetOption(lsUid, uid);
  return { cost, lsUid, type: constants.LORESHEET_BUY_OPTION, uid };
}
