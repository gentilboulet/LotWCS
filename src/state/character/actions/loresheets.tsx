import { createAction } from "typesafe-actions";

import * as dataLoresheets from "../../../data/loresheets";
import { ICost } from "../models/costs";

export const open = createAction(
  "loresheet/OPEN",
  action => (uid: string, cost: ICost) => {
    dataLoresheets.validateLoresheet(uid);
    return action({ cost, uid });
  }
);

export const buyOption = createAction(
  "loresheet/BUY_OPTION",
  action => (lsUid: string, uid: string, cost: ICost, payload?: string) => {
    dataLoresheets.validateLoresheetOption(lsUid, uid);
    return action({ cost, lsUid, uid, payload });
  }
);
