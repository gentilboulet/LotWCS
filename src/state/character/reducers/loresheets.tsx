import { produce, Draft } from "immer";
import { ActionType, getType } from "typesafe-actions";

import * as actions from "../actions/loresheets";
import { ILoresheetsState } from "../models/loresheets";

import { buyLoresheetOption, openLoresheet } from "../models//loresheets";

export const loresheetsReducer = produce(
  (draft: Draft<ILoresheetsState>, action: ActionType<typeof actions>) => {
    switch (action.type) {
      case getType(actions.open):
        openLoresheet(draft, action.payload.uid);
        break;
      case getType(actions.buyOption):
        buyLoresheetOption(
          draft,
          action.payload.lsUid,
          action.payload.uid,
          action.payload.payload,
        );
        break;
      default:
    }
  },
);
