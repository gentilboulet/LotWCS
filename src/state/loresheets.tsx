import { optionLoresheetData, validateLoresheet, validateLoresheetOption } from 'data/loresheets';

export interface ILoresheetOptionState {
  uid: string;
}

export interface ILoresheetsState {
  [lsUid: string]: ILoresheetOptionState[];
}

export function createState() : ILoresheetsState {
  return {};
}

export function openLoresheet(state: ILoresheetsState, loresheetUid: string): void {
  validateLoresheet(loresheetUid);
  state[loresheetUid] = [];
}

export function isLoresheetPresent(state: ILoresheetsState, loresheetUid: string) : boolean {
  const loresheetIndex = Object.keys(state).findIndex(stateLoresheetUid => stateLoresheetUid===loresheetUid);
  return loresheetIndex !== -1;
}

export function isLoresheetOptionPresent(state: ILoresheetsState, loresheetUid: string, optionUid: string) : boolean {
  const optionIndex = state[loresheetUid].findIndex(option => option.uid === optionUid);
  return optionIndex !== -1;
}

export function addLoresheetOption(state: ILoresheetsState, loresheetUid: string, optionUid: string): void {
  validateLoresheet(loresheetUid);
  validateLoresheetOption(loresheetUid, optionUid);

  if (! isLoresheetPresent(state, loresheetUid) ) { throw new Error('Internal error : loresheet not found'); }
  if ( isLoresheetOptionPresent(state, loresheetUid, optionUid)
  && !optionLoresheetData(loresheetUid, optionUid).repeatable) {
    throw new Error('Internal error : loresheet option is not repeatable');
  }
  /* TODO check prerequisites */

  state[loresheetUid].push({uid: optionUid});
}
