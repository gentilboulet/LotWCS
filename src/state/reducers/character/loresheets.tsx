import { Draft, produce } from "immer";
import { ActionType, getType } from "typesafe-actions";

import * as actions from "../../actions/character/loresheets";
import {
  buyLoresheetOption,
  ILoresheetsState,
  openLoresheet,
} from "../../models/character/loresheets";

export const loresheetsReducer = produce(
  (draft: Draft<ILoresheetsState>, action: ActionType<typeof actions>) => {
    switch (action.type) {
      case getType(actions.open):
        openLoresheet(draft, action.payload.uid);
        break;
      case getType(actions.buyOption):
        buyLoresheetOption(
          draft,
          action.payload.lsUid,
          action.payload.uid,
          action.payload.payload,
        );
        break;
      default:
    }
  },
);
