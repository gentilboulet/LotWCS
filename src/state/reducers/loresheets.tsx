import produce from "immer"

import { ILoresheetAction } from 'state/actions/loresheets';
import { IStoreState } from 'state/type';

import { addLoresheetOption, openLoresheet } from 'state/loresheets';

import * as constants from 'state/constants/loresheets';

export function loresheetsReducer(baseState: IStoreState, action: ILoresheetAction): IStoreState {
  switch (action.type) {
    case constants.LORESHEET_OPEN:
      return produce(baseState, draftState => {
        // applyCosts(draftState, action.cost)
        openLoresheet(draftState.loresheets, action.uid);
        draftState.history.push(action);
      });
    case constants.LORESHEET_BUY_OPTION:
      return produce(baseState, draftState => {
        // applyCosts(draftState, action.cost)
        addLoresheetOption(draftState.loresheets, action.lsUid, action.uid);
        draftState.history.push(action);
      });
    default:
  }
  return baseState;
}
