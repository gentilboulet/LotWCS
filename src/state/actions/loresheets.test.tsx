import * as actions from './loresheets';
import * as dataLoresheets from 'data/loresheets';

describe('Testing loresheet action creators', () => {
  const noCost = {
    destiny: 0, entanglement: 0,
    discountIdx: -1, discountNewValue: 0
  };

  it('should create an open action', () => {
    const lsUid = dataLoresheets.loresheets[0].uid;
    const a = actions.open(lsUid, noCost);
    expect( a.uid ).toBe(lsUid);
    expect( a ).toMatchSnapshot();
  });

  it('should not create an invalid open action', () => {
    const lsUid = 'invalid uid yeah';
    expect( () => actions.open(lsUid, noCost) ).toThrow();
  });

  it('should create a buy option action', () => {
    const lsUid = dataLoresheets.loresheets[0].uid;
    const optUid = dataLoresheets.loresheets[0].options[0].uid;
    const a = actions.buyOption(lsUid, optUid, noCost);
    expect( a.lsUid ).toBe(lsUid);
    expect( a.uid ).toBe(optUid);
    expect( a ).toMatchSnapshot();
  });

  it('should not create a buy option action for an invalid loresheet', () => {
    const lsUid = 'invalid uid yeah';
    const optUid = 'some uid, yeah';
    expect( () => actions.buyOption(lsUid, optUid, noCost) ).toThrow();
  });

  it('should not create a buy invalid option action', () => {
    const lsUid = dataLoresheets.loresheets[0].uid;
    const optUid = 'some uid, yeah';
    expect( () => actions.buyOption(lsUid, optUid, noCost) ).toThrow();
  });
});
