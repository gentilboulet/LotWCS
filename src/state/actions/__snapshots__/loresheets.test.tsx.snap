// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing loresheet action creators should create a buy option action 1`] = `
Object {
  "error": undefined,
  "meta": undefined,
  "payload": Object {
    "cost": Object {
      "canPay": true,
      "destiny": 0,
      "original": 0,
    },
    "lsUid": "jianghu",
    "payload": undefined,
    "uid": "familiarity",
  },
  "type": "loresheet/BUY_OPTION",
}
`;

exports[`Testing loresheet action creators should create an open action 1`] = `
Object {
  "error": undefined,
  "meta": undefined,
  "payload": Object {
    "cost": Object {
      "canPay": true,
      "destiny": 0,
      "original": 0,
    },
    "uid": "jianghu",
  },
  "type": "loresheet/OPEN",
}
`;
